lib/trino-parquet/src/main/java/io/trino/parquet/reader/flat/FlatColumnReader.java

```diff
@@ -370,4 +345,157 @@ private static void throwEndOfBatchException(int remainingInBatch)
     {
         throw new ParquetDecodingException(format("Corrupted Parquet file: extra %d values to be consumed when scanning current batch", remainingInBatch));
     }
+
+    /**
+     * A helper class which encapsulates reading data values from ValueDecoder into a buffer
+     * and creation of appropriate Block from the buffer values based on nulls count and presence of
+     * dictionary encoding.
+     */
+    private static class DataValuesBuffer<T>
+    {
+        private final ColumnAdapter<T> columnAdapter;
+        private final DictionaryDecoder<T> dictionaryValuesDecoder;
+        private final int batchSize;
+        @Nullable
+        private final T values;
+        @Nullable
+        private final int[] ids;
+        private int totalNullsCount;
+
+        private DataValuesBuffer(
+                ColumnAdapter<T> columnAdapter,
+                DictionaryDecoder<T> dictionaryValuesDecoder,
+                boolean hasOnlyDictionaryEncodedPages,
+                int batchSize)
+        {
+            this.columnAdapter = requireNonNull(columnAdapter, "columnAdapter is null");
+            this.batchSize = batchSize;
+            if (hasOnlyDictionaryEncodedPages) {
+                this.dictionaryValuesDecoder = requireNonNull(dictionaryValuesDecoder, "dictionaryValuesDecoder is null");
+                this.ids = new int[batchSize];
+                this.values = null;
+            }
+            else {
+                this.dictionaryValuesDecoder = null;
```

Comment:
can there be:
```
dictionary encoded page
plain page
dictionary encoded page
```
?

Reply:
Fallback is one-way in parquet-mr writer.
If some other writer did that, it would need to always keep track of whether all the entries in a page fall within the dictionary even after fallback to non-dictionary encoding. It's unlikely that would be a good trade-off.
If it did happen, we would detect that through `pageReader.hasOnlyDictionaryEncodedPages()` and go to the dictionary resolving code path which won't be affected by this.